### DISCUSSION TITLE
Discussion: GraphQL Spec Compliance Tests

### DISCUSSION THREAD

**Discussion by pranshuchittora** [graphql-java/graphql-java#2477] [General]
**Edit**

##### Spec discussion  -> https://github.com/graphql/graphql-spec/issues/698
---
Hello maintainers
As a part of research around the simplifying graphql spec tests across implementations, I've stumbled upon the specs written here for `graphql-java`. I'd really appreciate if I can get an idea around how difficult it has been to maintain this test suite, and how often does it go out of parity with the actual graphql spec? Were there cases when misunderstanding the spec led to issues with the library? 

#### Would having a generic set of compatibility tests help around these issues if any?

As schema definition in SDL format is natively supported by the library then it would be easy to adapt generic testing suite across different languages.

**Comment by andimarek** [comment]
Hi,
Can you give us a bit more information about what kind of research you are doing and for whom?

An automated test suite is something all maintainers would love to have. But it is a serious effort and involves a couple
challenging details. The best effort so far was this project: https://github.com/graphql-cats/graphql-cats

**Comment by bbakerman** [comment]
For more detail on automated test for graphql implementations the `cats` are a good first attempt however its the low level details that stop this on practice.

For example lets say we want to have a test for "Fragements must not be circular".

https://spec.graphql.org/June2018/#sec-Fragment-spreads-must-not-form-cycles

Specifying this in cats is straight foward ish however how to `assert` was the challenge

We might have an error message with the text "Circular fragments not allowed" and grpahql-js might have "You are not allowed to make circular fragments"

Now each implementation of cats requires you to map a logical test to your own errors.  This greatly reduces the "cats" value and dramtically increased the cost to be 'cats' ready.

It got worse in other test cases.

Ideally we could have common identifiers in errors that cats could know to look for say.  Imagine `error.extensions.specList = [5.5.2.2]` which is the number of that rule in the spec.

The initial `cats` tests where too specific to the Scala implementation - it was a great first step in terms of ideas.  It got a little better but not by enough so we stop trying to follow it - the value just want not there.

That said we do not have a great way to (quasi) formally prove that we are spec compliant on query execution and graphql type system SDL generation.

**Comment by pranshuchittora** [comment]
> Hi,
> Can you give us a bit more information about what kind of research you are doing and for whom?

Actually I am researching for the project proposal for the project idea given by `GraphQL Foundation` for `Google Summer of Code 2020` -> [Project Idea](https://github.com/graphql/foundation/tree/master/mentorship/2020/gsoc#1-graphql-compatibility-acceptance-tests-medium) 

The idea is to implement a `Compatibility Acceptance Test Suite` which can be integrated into the GraphQL implementation library.

**Comment by pranshuchittora** [comment]
@bbakerman I completely understand that integrating `cats` requires a lot of effort. 
Hence there are always some trade-offs, to make the testing more generic the ideal way is to test for only the input-output.
This can be done by spawning an instance of GraphQL server (graphql-java) and test it on various queries & schemas over HTTP. As all the implementations support running over HTTP.

@ maintainers how does that sound in terms of maintainability and community acceptance? Is there any alternative design that you would like to suggest?

**Comment by pranshuchittora** [comment]
@andimarek the `cats` project was not easy to integrate, as it's been designed to be coupled with the library's testing infra. My idea is to make it decoupled & more of a plug-and-play kind of thing so that it can be easy to integrate and disintegrate.

As mentioned above, the GraphQL server will be tested over HTTP with a client provided. The client will send requests and evaluates the responses.

**Comment by bbakerman** [comment]
Yeah so if you can come up with a more concrete proposal about how this would work and so on then we would be receptive.

Again I urge you to consider more deeply how to identify errors and conditions - these are the low level blockers that halted the other effort.

For example we can run up a HTTP graphql server (given a schema) and you could fire off a request with a circular fragment in it- it will error - BUT how do you know it errored on that condition and only that condition?

How can you look at the response and know its a circular fragment  error  and not something else.

One could argue that's enough but once you get into this I suspect you will find that you need to know more detailed information for assertions


All that said I can imagine such a system requiring

* a series of schemas that it must accept - it you POST in a schema and the server graphql tester makes up a server for that schema
*   POST /useschema/{testname}
* POST /graphql/{testname} # for queries on that server instance


* a based data set that fetchers should give out for a test case

* POST /usedata/{testname}
* POST /graphql/{testname} # for queries on that server instance using that data

**Comment by pranshuchittora** [comment]
@bbakerman thanks for the awesome feedback :)
> Yeah so if you can come up with a more concrete proposal about how this would work and so on then we would be receptive.

Actually I am researching and asking all maintainers of the other libraries as well to get to some robust design because at the end of the day it's going to be integrated by the approval of the maintainers only. Hence requirement analysis is required to eliminate any bad approach.


>For example we can run up a HTTP graphql server (given a schema) and you could fire off a request with a circular fragment in it- it will error - BUT how do you know it errored on that condition and only that condition?

As far as I know, the spec doesn't describe any specific error message, therefore, various implementations have different error strings. Correct me if I am wrong.
In future, the committee might come with some RFC to address this issue by either generalising the error message or HTTP codes. :tipping_hand_man: 

> a series of schemas that it must accept - it you POST in a schema and the server graphql tester makes up a server for that schema

Yeah, you are correct. I am thinking of a master HTTP server with a couple of REST APIs exposed. This server will listen for events on various endpoints from the GraphQL client. Like
- Update the Schema/fakeData  (Given schema/fakeData as payload)
- Spawn the GraphQL server with the updated schema
- Restart, kill the GraphQL instance

The client can run in either CLI (for running test in CI) or on a browser (Given master is reachable).

**Comment by bbakerman** [comment]
> As far as I know, the spec doesn't describe any specific error message, therefore, various implementations have different error strings. Correct me if I am wrong.

Yup

> In future, the committee might come with some RFC to address this issue by either generalising the error message or HTTP codes. 💁‍♂

Has been talked about before (along with cats) but it got no traction - but I hold out hope

For example imagine a well known error place that can hold a "spec rule number" that is indicative of what validation rule you have caused an error message to fail on

` errors : [ { message : "any old message", extensions : [ __graphqlspec : { reason : "March2018-5.1.2.4" }}}}`


> The client can run in either CLI (for running test in CI) or on a browser (Given master is reachable).

I think its ok for a CLI client to HTTP to a server even if its on a local machine

**Comment by pranshuchittora** [comment]
@bbakerman the error approach you mentioned is somewhat relatable to the `rust` error handling i.e. the spec has defined the errors and their respective codes.
https://doc.rust-lang.org/error-index.html

Similarly, the TypeScript also comes with error codes for various errors.
https://github.com/microsoft/TypeScript/blob/0aa2e2783c42dd48e7d3085e1a612ac410416ec1/src/compiler/diagnosticMessages.json

If GraphQL spec comes with some sort of error codes then it would be great for users too. As users can directly google the error codes and read the error definition from the official source.

**Comment by pranshuchittora** [comment]
Is there any RFC regarding the error codes?

**Comment by bbakerman** [comment]
Not that I know of.

hey rather than an issue lets start a discussion for this in specturm

https://spectrum.chat/graphql-java?tab=posts

**Comment by pranshuchittora** [comment]
> Not that I know of.
> 
> hey rather than an issue lets start a discussion for this in specturm
> 
> https://spectrum.chat/graphql-java?tab=posts

That sounds great. Any GraphQL foundation members are there?
